Source/Serilog.Exceptions/Reflection/ReflectionInfoExtractor.cs

```diff
@@ -0,0 +1,122 @@
+namespace Serilog.Exceptions.Reflection
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Linq.Expressions;
+    using System.Reflection;
+
+    /// <summary>
+    /// Utility class that analyzes type using reflection and provides
+    /// information about properties to be used in destructuring process.
+    /// </summary>
+    internal class ReflectionInfoExtractor
+    {
+        private readonly object lockObj = new();
+        private readonly Dictionary<Type, ReflectionInfo> reflectionInfoCache = new();
```

Comment:
Could we use a `ConcurrentDictionary` instead and ditch the `lock`?

Reply:
I will save that idea and come back to it afterward. Surely I do not want mix such a change to this PR.